### PR TITLE
refactor: Rework undo to typescript

### LINKED ISSUE #44433: convert redux/undo and UndoListing.jsx to TS

### PR DESCRIPTION
**Pull request by uladzimirdev**
Closes https://github.com/metabase/metabase/issues/44433

### Description

Add types to redux/undo and UndoListing

### How to verify

- ci is green
- types are correctly specified

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Review by kamilmielnik — COMMENTED** [review]
Nice! :medal_military:

File: frontend/src/metabase/redux/undo.ts [hunk 1]
```diff
@@ -93,34 +101,35 @@ export const getIsRecentlyAutoConnectedDashcard = createSelector(
...
+      undo.actions?.map(action => dispatch(action));
+      dispatch(dismissUndo({ undoId, track: false }));
     }
   };
 });
 
-export default function (state = [], { type, payload, error }) {
+export function undoReducer(
+  state: Undo[] = [],
+  { type, payload, error }: Action<Undo>,
```
[the remark this answers is not in the document]
Comment (uladzimirdev) [reply frontend/src/metabase/redux/undo.ts:131]: I tried it as well, good point. I came up to the idea that we'll need to rework this reducer to RTK anyway and `builder.addCase` supports custom payload type, so I decided not to waste time on this check

**Review by ranquild — APPROVED** [review]

File: frontend/src/metabase/redux/undo.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (uladzimirdev) [reply frontend/src/metabase/redux/undo.ts]: [5489cac](https://github.com/metabase/metabase/pull/44479/commits/5489cac71fd13faba8320be785943fb629da18e9)

File: frontend/src/metabase/redux/undo.ts [hunk 2]
```diff
@@ -202,7 +214,7 @@ export default function (state = [], { type, payload, error }) {
   return state;
 }
 
-const clearTimeoutForUndo = undo => {
+const clearTimeoutForUndo = (undo: Undo) => {
   if (undo?.timeoutId) {
```
[the remark this answers is not in the document]
Comment (uladzimirdev) [reply frontend/src/metabase/redux/undo.ts]: [5489cac](https://github.com/metabase/metabase/pull/44479/commits/5489cac71fd13faba8320be785943fb629da18e9)

File: frontend/src/metabase/redux/undo.ts [hunk 3]
```diff
@@ -36,8 +45,10 @@ export const addUndo = createThunkAction(ADD_UNDO, undo => {
 });
 
 const PAUSE_UNDO = "metabase/questions/PAUSE_UNDO";
-export const pauseUndo = createAction(PAUSE_UNDO, undo => {
-  clearTimeout(undo.timeoutId);
+export const pauseUndo = createAction(PAUSE_UNDO, (undo: Undo) => {
+  if (undo.timeoutId) {
+    clearTimeout(undo.timeoutId);
+  }
```
[the remark this answers is not in the document]
Comment (uladzimirdev) [reply frontend/src/metabase/redux/undo.ts]: [5489cac](https://github.com/metabase/metabase/pull/44479/commits/5489cac71fd13faba8320be785943fb629da18e9)

File: frontend/src/metabase/redux/undo.ts [hunk 4]
```diff
@@ -16,12 +19,18 @@ export const addUndo = createThunkAction(ADD_UNDO, undo => {
     const { icon = "check", timeout = 5000, canDismiss = true } = undo;
     const id = undo.id ?? nextUndoId++;
     // if we're overwriting an existing undo, clear its timeout
-    const currentUndo = getUndo(getState(), id);
-    clearTimeoutForUndo(currentUndo);
+    const currentUndo = getUndo(getState().undo, id);
```
[the remark this answers is not in the document]
Comment (uladzimirdev) [reply frontend/src/metabase/redux/undo.ts]: [5489cac](https://github.com/metabase/metabase/pull/44479/commits/5489cac71fd13faba8320be785943fb629da18e9)

**Review by kamilmielnik — APPROVED** [review]

**Pull request merged**